In [1]:
#-*- coding: utf-8 -*-

# Analiza tekstu

## Zadanie

Napiszemy program, który czyta utwór literacki z pliku tekstowego i następnie dokonamy analizy tego utworu.

# Krok 1: Wczytanie nazwy pliku

Do czytania danych ze standardowego wejścia służy funkcja `input()`:
```python
input()
input('Podaj nazwę pliku')
```

In [ ]:
input()

# Krok 2: Sprawdzenie, czy plik istnieje

Najpierw musimy zaimportować moduł

```python
import os
```
Dzięki temu modułowi możemy sprawdzić, czy plik istnieje:
```python
os.path.exists(<ścieżka_do_pliku>)
```


Metoda zwraca wartość logiczną. Możemy więc skorzystać z instrukcji warunkowej:
   
```python
if warunek_logiczny_1:
    # instrukcje wykonywane, gdy warunek_logiczny_1 jest prawdziwy
elif warunek_logiczny_2:
    # instrukcje wykonywane, gdy warunek_logiczny_1 nie jest prawdziwy,
    # ale warunek_loginczy_2 jest prawdziwy
else:
    # instrukcje wykonywane, gdy ani warunek_logiczny_1 ani
    # warunek_logiczny_2 nie zachodzą
```

## Podsumowanie

Możemy napisać kod:

In [ ]:
import os

file_path = input("Podaj ścieżkę do pliku: ")
if os.path.exists(file_path):
    print("Plik " + file_path + " istnieje")
else:
    print("Plik " + file_path + " NIE istnieje")

# Krok 3: Wymuszenie podania właściwego pliku

Program, w którym nie będziemy mieli ścieżki do istniejącego pliku, jest bezużyteczny. Musimy wymusić na użytkowniku podanie właściwej ścieżki. Możemy to zrobić przy użyciu pętli:

```python
while warunek:
    # instrukcje wykonywane dopóki warunek jest prawdziwy
```

## Podsumowanie

Poprzedni kod możemy przepisać:

In [ ]:
import os

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

# Krok 4: Czytanie pliku tekstowego

Do otwarcia pliku do czytania używamy instrukcji:
```python
f = open(<ścieżka do pliku>)
```
A gdy chcemy pisać do pliku:
```python
f = open(<ścieżka do pliku>, 'w')
```

## Sposób 1:

Możemy przeczytać cały plik do jednej zmiennej:

In [8]:
f = open('reduta-ordona.txt')
content = f.read()
print(r)

## Sposób 2:

Możemy czytać plik linijka po linijce:

In [18]:
f = open('reduta-ordona.txt')
line = f.readline()
while line != '':
    print(line)
    line = f.readline()

## Sposób 3:

Możemy przeczytać plik do listy linijek:

In [19]:
f = open('reduta-ordona.txt')
lines = f.readlines()

## Listy w pythonie

### Tworzenie listy:

In [22]:
l = [1, 2, 3.0, 'pawel', True]
print(l)

[1, 2, 3.0, 'pawel', True]


### Dostęp do dowolnego elementu w liście używając indeksu:

In [24]:
print(l[0])
print(l[3])
print(l[-1])

1
pawel
True


### Wybór podlisty:

In [25]:
print(l[:3])
print(l[2:])
print(l[1:-1])

[1, 2, 3.0]
[3.0, 'pawel', True]
[2, 3.0, 'pawel']


### Znalezienie długości listy

In [30]:
len(l)

5

### Powiększanie listy

In [40]:
print(l)
l.append(4)
print(l)
print(l + [3,2,1])

[1, 2, 3.0, 'pawel', True, 4, 4]
[1, 2, 3.0, 'pawel', True, 4, 4, 4]
[1, 2, 3.0, 'pawel', True, 4, 4, 4, 3, 2, 1]


### Sprawdzenie, czy element jest na liście

In [41]:
print(1 in l)
print('bogdan' in l)

True
False


### Przeglądanie całej listy

In [26]:
for element in l:
    print(element)

1
2
3.0
pawel
True


## Uwaga

Wszystko powyższe stosuje się do napisów. Napis należy rozumieć, jako listę liter

In [28]:
napis = "Paweł Bogdan"
print(napis)
print(napis[0])
print(napis[-1])
print(napis[:5])

Paweł Bogdan
P
n
Paweł


## Pętla `for each`

```python
for element in lista:
    # w każdym przebiegu pętli
    # w zmiennej element będzie kolejny element znajdujący się na liście
```

# Krok 5: policzenie częstotliwości występowania poszczególnych liter

Możemy skorzystać z funkcji `ord`:

In [33]:
print(ord('a'))
print(ord('z'))
print(ord('A'))
print(ord('Z'))

97
122
65
90


Tworzymy wtedy tablicę o rozmiarze 122 i jesteśmy w stanie w odpowiedniej komórce przechowywać ilość wystąpień liter o zadanych kodach ASCII.

## Wady?

### Dużo nieużywanych komórek, które zajmują miejsce w pamięci

### A co z innymi literami?

In [35]:
print(ord('ą'))
print(ord('Ź'))

261
377


### Potrzebujemy tablicy, w której indeksami może być coś innego niż kolejne liczby naturalne

## Rozwiązanie: słownik

### Tworzenie pustego słownika

In [36]:
d = {}

### Tworzenie słownika

In [37]:
d = {'a': 1, 'v': 5, 6: 'pawel', 4.5: 2}

### Dostęp do wartości:

In [43]:
print(d['a'])
print(d[6])

1
pawel


### Edycja słownika

In [44]:
d['a'] = 4
d['bogdan'] = 'pawel'

### Przeglądanie słownika

In [53]:
print(d)
print(d.keys())
print(d.values())
for k in d:
    print(d[k])
print(d.items())
print('a' in d)

{'a': 4, 'v': 5, 6: 'pawel', 4.5: 2, 'bogdan': 'pawel'}
dict_keys(['a', 'v', 6, 4.5, 'bogdan'])
dict_values([4, 5, 'pawel', 2, 'pawel'])
4
5
pawel
2
pawel
dict_items([('a', 4), ('v', 5), (6, 'pawel'), (4.5, 2), ('bogdan', 'pawel')])
True


## Dygresja - krotki

### Lista:

In [50]:
l = [2,3,4]
print(l)

[2, 3, 4]


### Krotka

(po angielsku: _tuple_)

In [51]:
k = (2,3,4)
print(k)

(2, 3, 4)


### Przeglądanie listy krotek

In [52]:
for k, v in d.items():
    print(k,v)

a 4
v 5
6 pawel
4.5 2
bogdan pawel


## Podsumowanie

In [55]:
import os

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

f = open(file_path)
freq = {}
for letter in f.read():
    if not (letter in freq):
        freq[letter] = 1
    else:
        freq[letter] = freq[letter] + 1

for k, v in freq.items():
    print("{0} -> {1}".format(k, v))

Podaj ścieżkę do pliku: reduta-ordona.txt


# Krok 6: Wyeliminowanie niepotrzebnych znaków

Do statystyki liter nie są nam potrzebne białe znaki, znaki przystankowe i cyfry.

Do sprawdzenia, czy napis w zmiennej `s` zawiera tylko litery możemy wykorzystać wywołanie: `s.isalpha()`.


## Podsumowanie

In [ ]:
import os

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

f = open(file_path)
freq = {}
for letter in f.read():
    if letter.isalpha():
        if not (letter in freq):
            freq[letter] = 1
        else:
            freq[letter] = freq[letter] + 1

for k, v in freq.items():
    print("{0} -> {1}".format(k, v))

## Albo

In [ ]:
import os

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

f = open(file_path)
freq = {}
for letter in f.read():
    if not letter.isalpha():
        continue
    if not (letter in freq):
        freq[letter] = 1
    else:
        freq[letter] = freq[letter] + 1

for k, v in freq.items():
    print("{0} -> {1}".format(k, v))

# Krok 7: Brak rozróżnienia, czy litera jest wielka czy mała

Do sprawdzenia, czy napis w zmiennej `s` składa się tylko z wielkich liter można użyć polecenia `s.isupper()`, a czy składa się tylko z małych liter polecenia `s.islower()`.

Możemy też znaleźć odpowiednik napisu w zmiennej `s` w którym wszystkie litery są zamienione na wielkie przez użycie `s.upper()` i podobnie z małymi literami: `s.lower()`.

## Podsumowanie

In [7]:
import os

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

f = open(file_path)
freq = {}
for letter in f.read():
    if not letter.isalpha():
        continue
    l = letter.lower()
    if not (l in freq):
        freq[l] = 1
    else:
        freq[l] = freq[l] + 1

for k, v in freq.items():
    print("{0} -> {1}".format(k, v))

Podaj ścieżkę do pliku: reduta-ordona.txt


# Krok 8: Posortowanie wyników

Każdą listę `l` możemy posortować używając polecenia `l.sort()`. Instrukcja nie zwraca wartości, ale sprawia, że lista `l` będzie posortowana. Można też użyć funkcji `sorted(l)`, która nie zmienia listy `l`, ale zwraca posortowaną kopię.

## Podsumowanie

In [8]:
import os

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

f = open(file_path)
freq = {}
for letter in f.read():
    if not letter.isalpha():
        continue
    l = letter.lower()
    if not (l in freq):
        freq[l] = 1
    else:
        freq[l] = freq[l] + 1
for k, v in sorted(freq.items()):
    print("{0} -> {1}".format(k, v))

Podaj ścieżkę do pliku: reduta-ordona.txt


# Po co badać częstotliwość liter w tekście?

# Cofnijmy się o 2000 lat

![Juliusz Cezar](https://vignette.wikia.nocookie.net/obelix-i-asterix/images/6/63/Juliusz-Cezar.png/revision/latest?cb=20141103153822&format=original&path-prefix=pl)
[Źródło](http://pl.obelix-i-asterix.wikia.com/wiki/Juliusz_Cezar)

# Szyfr Cezara
![Szyfr Cezara](https://upload.wikimedia.org/wikipedia/commons/thumb/2/2b/Caesar3.svg/500px-Caesar3.svg.png)
[Źródło](https://commons.wikimedia.org/wiki/File:Caesar3.svg)

# Szyfr Cezara w Pythonie

## Co będzie potrzebne?

1. Alfabet szyfrowanego tekstu
2. Znalezienie, którą pozycję w alfabecie zajmuje dana litera
3. Znalezienie pozycji, którą w alfabecie zajmuje litera, którą mamy wstawić do tekstu zaszyfrowanego
4. Znalezienie właściwej litery wyjściowej

# Funkcje

```python
def nazwa_funkcji(argument_1, argument_2, ..., argument_n):
    ## Ciało funkcji
    ## Jeżeli chcemy, żebu funkcja zwracała jakąś wartość:
    return value
```

## Alfabet

In [13]:
alphabet = ['a', 'ą', 'b', 'c', 'ć', 'd', 'e', 'ę', 'f', 'g', 
            'h', 'i', 'j', 'k', 'l', 'ł', 'm', 'n', 'o', 'ó', 
            'p', 'r', 's', 'ś', 't', 'u', 'w', 'y', 'z', 'ż', 'ź']

## Szukanie pozycji litery w alfabecie

Chcemy mieć funkcję `znajdz_pozycje`, która przyjmie jako argument alfabet i literę a zwróci pozycję zajmowaną przez literę w alfabecie

In [14]:
def znajdz_pozycje(alfabet, litera):
    for i in range(len(alfabet)):
        if alfabet[i] == litera:
            return i

## Szukanie litery szyfrującej

Potrzebujemy funkcji `zaszyfruj_litere`, która przyjmie alfabet i literę do zaszyfrowania, a zwróci literę szyfrującą podaną na wejściu.

In [15]:
def zaszyfruj_litere(alfabet, litera):
    pozycja = znajdz_pozycje(alfabet, litera)
    nowa_pozycja = (pozycja + 3) % len(alfabet)
    return alfabet[nowa_pozycja]

## Szyfrowanie całego napisu

Teraz chcemy napisać funkcję `zaszyfruj`, która jako argument przyjmie alfabet i cały napis, który chcemy zaszyfrować, a jej wynikiem będzie zaszyfrowany napis

In [16]:
def zaszyfruj(alfabet, napis):
    wynik = ''
    for litera in napis:
        wynik = wynik + zaszyfruj_litere(alfabet, litera)
    return wynik

## Zadanie

Do tej pory robiliśmy statystykę częstotliwości liter w pobranym utworze. Teraz stwórzymy plik, który będzie zawierał zaszyfrowaną wersję naszego utworu.
__UWAGA__: Jeżeli jakiegoś znaku nie da się zaszyfrować, to go  wstawiamy do tekstu wynikowego bez zmiany.

## Rozwiązanie


In [18]:
def znajdz_pozycje(alfabet, litera):
    for i in range(len(alfabet)):
        if alfabet[i] == litera:
            return i
    return -1

def zaszyfruj_litere(alfabet, litera):
    pozycja = znajdz_pozycje(alfabet, litera)
    if pozycja == -1:
        return litera
    else:
        nowa_pozycja = (pozycja + 3) % len(alfabet)
        return alfabet[nowa_pozycja]

def zaszyfruj(alfabet, napis):
    wynik = ''
    for litera in napis:
        wynik = wynik + zaszyfruj_litere(alfabet, litera)
    return wynik

file_path = input("Podaj ścieżkę do pliku: ")

while not os.path.exists(file_path):
    file_path = input("Plik nie istnieje, spróbuj jeszcze raz: ")

f = open(file_path)
szyfr = zaszyfruj(alphabet, f.read())
f = open("zaszyfrowany.txt", 'w')
f.write(szyfr)
f.flush()
f.close()

Podaj ścieżkę do pliku: reduta-ordona.txt


# Szyfry podstawieniowe

$$\sigma~:~\Sigma \rightarrow \Sigma$$

|a          |ą          |b          |c          |ć          |...|y          |z          |ż          |ź          |
|:---------:|:---------:|:---------:|:---------:|:---------:|:-:|:---------:|:---------:|:---------:|:---------:|
|b|g|h|u|t|...|m|p|r|z|

## Analiza statystyczna

Sprawdźmy, co wypisze nasz program do zliczania częstości liter w tekście zaszyfrowanym

## Kryptoanaliza

![histogram](https://upload.wikimedia.org/wikipedia/commons/thumb/c/cd/Polish_letters_frequencies.svg/500px-Polish_letters_frequencies.svg.png)
[Źródło](https://commons.wikimedia.org/wiki/File:Polish_letters_frequencies.svg)

# Szyfry polialfabetyczne

## Szyfr Vigenere'a (le chiffre indéchiffrable)

![szyfr_Vigenerea](https://upload.wikimedia.org/wikipedia/commons/thumb/9/9a/Vigen%C3%A8re_square_shading.svg/600px-Vigen%C3%A8re_square_shading.svg.png)
[Źródło](https://commons.wikimedia.org/wiki/File:Vigen%C3%A8re_square_shading.svg)

# Enigma

![enigma1](https://upload.wikimedia.org/wikipedia/commons/a/ae/Enigma.jpg)
[Źródło](https://commons.wikimedia.org/wiki/File:Enigma.jpg)
[Symulator](http://enigmaco.de/enigma/enigma_pl.html)